Search refill optimisation

sim590 opened the pull request:
In order to optimize the refill of a search with good `SearchNode`, the refill is now done at each call of `searchStep()`. A refill will try to add more good nodes to the search until the search has reached `SEARCH_NODES` good nodes.

- [x] Include a notion of delay between two `Search::refill` calls to avoid network flood.


@aberaud: Please let know what you think. I'll squash some of the commits before you merge this when it is acceptable.

aberaud commented on src/dht.cpp:
```diff
@@ -1048,40 +1059,31 @@ Dht::searchStep(Search& sr)
+            // no nodes or all expired nodes
+            sr.expired = true;
+            // reset refilled since the search is now expired.
+            sr.refilled = false;
```
If the refill succeeds, then refilled will be true. If a second is needed later, it would not happen.

sim590 replied:
Indeed. I think we can even get rid of `Search::refilled` member. However, there should be some delay between two refills in order to avoid traffic catastrophies.

aberaud commented on src/dht.cpp:
```diff
@@ -939,6 +944,12 @@ Dht::searchStep(Search& sr)
     /* Check if the first TARGET_NODES (8) live nodes have replied. */
     if (sr.isSynced(now)) {
+
+        if (sr.nodes.size()-sr.getNumberOfCandidates(now) < SEARCH_NODES and not sr.refilled) {
```
Why restricting this to the synced state ? Synced means all 8 first known nodes have answered, so in that case there would be no need to refill at all.

sim590 replied:
Yes. You are right. I'll fix that.

sim590 commented:
I've included a delay between two successive refills. However, a stronger empirical analysis should be done around the accurate value to use.

@aberaud: What do you think?

sim590 commented:
This change will be included in a PR for #42. Closing this PR.

sim590 commented:
I think this should be merged outside of #52 since it doesn't have to do with refactoring... However, I have to reopen the PR since [I closed and changed rebased in the mean time][]

[I closed and changed rebased in the mean time]: https://github.com/isaacs/github/issues/361
